### ISSUE TITLE
Default Codespaces "universal" image is now based on Ubuntu 20.04

### ISSUE THREAD

**Issue by Chuxel** [microsoft/vscode-dev-containers#675]
## Ubuntu 20.04 (Focal) update for default GitHub Codespaces image

The `codespaces-linux` definition in this repository used to create the default "universal" image for GitHub Codespaces is updating from Debian 9 (Stretch) to Ubuntu 20.04 (Focal). Debian 9 is still in long-term-support, but was released in 2017 and many libraries are outdated particularly compared to Ubuntu 20.04 which was released this year. You will see a notice the first time you open a terminal with this update once service has switched to the update over the next week.

However, this new Ubuntu 20.04 image brings with it some potentially breaking changes due to native library updates and available / default packages. For example, it:

1. Defaults the `python` command to version 3.x (and you instead use `python2` to access the out-of-support version)
2. Drops older versions of certain libraries like `libssl1.0.2` that can break compatibility.
3. And as a result, required older SDKs/Runtimes to be dropped from the image (e.g. .NET Core 1, old versions of Node).

Because of these potentially unexpected impacts, we've increased the major version of the image so you can still access the old one (version 1.x.x is Focal while 0.x.x is Stretch).

> **Tip:** If needed, you can switch the default `python` command back to Python 2.x by simply running the following from a terminal in your codespace: `sudo apt-get install -y python-is-python2`

### Existing custom images referencing the Codespaces image

If you have an existing reference to `mcr.microsoft.com/vscode/devcontainers/universal:0-linux` in a custom Dockerfile or devcontainer.json, it will continue to reference the Debian 9 (Stretch) version.  However, note that **referencing this older version of the image will increase codespace create time** since it will no longer be cached. (That said, resuming an existing codespace is **unaffected**.) 

We recommend moving to `mcr.microsoft.com/vscode/devcontainers/universal:1-linux` as soon as possible or using a smaller custom image. If you'd prefer to always stay on latest version even after OS updates, you can reference `mcr.microsoft.com/vscode/devcontainers/universal:latest` instead.

### Continuing to use the older "Stretch" version of the image

If for some reason you need to use something in the old Debian 9 (Stretch) based image, you can add the following `.devcontainer/devcontainer.json` file to your repository before creating a codespace:

```json
{
    "image":"mcr.microsoft.com/vscode/devcontainers/universal:0-stretch",
    "settings": {
        "terminal.integrated.shell.linux": "/bin/bash",
    },
    "remoteUser": "codespace",
    "overrideCommand": false,
    "mounts": [ "source=/var/run/docker.sock,target=/var/run/docker-host.sock,type=bind" ],
    "runArgs": [ "--cap-add=SYS_PTRACE", "--security-opt", "seccomp=unconfined" ],
    "postCreateCommand": "oryx build  || echo 'Could not auto-build. Skipping.'"
} 
```

### Getting started with a custom image

This repository contains a number of starter custom dev containers. As a convenience, when inside a GitHub Codespaces or using the Remote - Containers extension locally, you can use the **Add Development Container Configuration Files...** command from the command palette (<kbd>F1</kbd>) to add one of these start custom container to your project. 

You can then commit and push this to source control. In GitHub Codespaces, you will need to create a new codespace for the setting to take effect as "rebuild" support is not yet available (though this support is planned).

**Comment by MoneyMan573** [comment]
{
    "image":"mcr.microsoft.com/vscode/devcontainers/universal:0-stretch",
    "settings": {
        "terminal.integrated.shell.linux": "/bin/bash",
    },
    "remoteUser": "codespace",
    "overrideCommand": false,
    "mounts": [ "source=/var/run/docker.sock,target=/var/run/docker-host.sock,type=bind" ],
    "runArgs": [ "--cap-add=SYS_PTRACE", "--security-opt", "seccomp=unconfined" ],
    "postCreateCommand": "oryx build  || echo 'Could not auto-build. Skipping.'"
}

**Comment by berkant** [comment]
Is `linux` tag any different than `focal`? I don't really get why there is the `linux` tag in the first place.

**Comment by Chuxel** [comment]
The `1` is a partial version number that prevents breaking changes from being picked up.  The currently released version is `1.3.3` so you can pick up `1`, `1.3` or `1.3.3` to get it.  However, if tomorrow we release `1.3.4`, you'd get that with `1` or `1.3`.  This is [described in the readme](https://github.com/microsoft/vscode-dev-containers/tree/master/containers/codespaces-linux)

You can also check out the history file for the specific tags that are available for a given version: [1.3.3](https://github.com/microsoft/vscode-dev-containers/blob/master/containers/codespaces-linux/history/1.3.3.md)

The `-linux` prefix is because there was discussions of a Windows container version - though that is not currently in plan.  `-focal` distinguishes it from `-stretch` which was the old Debian 9 version.
